security/s2a/internal/crypter/halfconn.go

```diff
@@ -0,0 +1,126 @@
+package crypter
+
+import (
+	"fmt"
+	"golang.org/x/crypto/cryptobyte"
+	s2a_proto "google.golang.org/grpc/security/s2a/internal"
+	"hash"
+	"sync"
+)
+
+const (
+	tls13Key    = "tls13 key"
+	tls13Nonce  = "tls13 iv"
+	tls13Update = "tls13 upd"
+)
+
+type S2AHalfConnection struct {
+	h             func() hash.Hash
+	aeadCrypter   s2aAeadCrypter
+	expander      hkdfExpander
+	seqCounter    counter
+	mutex         sync.Mutex
+	trafficSecret []byte
+	nonce         []byte
+}
+
+func NewHalfConn(ciphersuite s2a_proto.Ciphersuite, trafficSecret []byte) (S2AHalfConnection, error) {
+	cs := NewCiphersuite(ciphersuite)
+	if cs.trafficSecretSize() != len(trafficSecret) {
+		return S2AHalfConnection{}, fmt.Errorf("supplied traffic secret must be %v bytes, given: %v", cs.trafficSecretSize(), trafficSecret)
+	}
+
+	hc := S2AHalfConnection{h: cs.hashFunction(), expander: &defaultHKDFExpander{}, seqCounter: newCounter()}
```

Comment:
Can we move this line after we derive the key and the nonce? That way, if an error occurs while deriving the key/nonce then we don't create the half connection at all.

Reply:
I think you mean "then we don't create an AEAD crypter at all" correct? Since we don't make a half connection at all if there's a failure anywhere at all.

Reply:
But doesn't the `hc := S2AHalfConnection(...)` line initialize the half connection? (Modulo some variables not being set.)

In any event, I think it's best to fail fast, i.e. derive the key/nonce sooner (since those operations can fail) and then initialize `hc`.

Reply:
Oh I see what you're saying, I think I misunderstood your first comment. `hc := S2AHalfConnection(...)` initializes the half connection but an empty half connection is returned on any error. We still need the half connection before we call `deriveSecret` since `deriveSecret` is a method of the half connection. I've removed the hash function from the parameter of `deriveSecret` just now since it has access to the hash function from the half connection anyways.

Reply:
If we want to move `hc := S2AHalfConnection(...)` after we derive the secrets, we'll have to include the hash function as well as the expander in the `deriveSecret` API. I think it would be cleaner just to initialize the half connection first and then call `deriveSecret` after. The cost to initializing the half connection is minimal anyways, and it makes the code cleaner in my opinion (due to the cleaner `deriveSecret` parameters). WDYT?